### PR TITLE
chore: Made a single scripts for all functions to be run in travis.

### LINKED ISSUE #2841: Create a single script for all functions to be run doing the travis build.
**Describe the chore**
Create a single script for all functions to be run doing the travis build.

**Would you like to work on the issue?**

- [X] Yes
- [ ] No
- Other:

### PR DESCRIPTION
**Pull request by yashk2000**
Fixes #2841 

Made a single scripts for all functions to be run in travis.

### CODE REVIEW

**Comment by iamareebjamal** [comment]
This is opposite of what I asked. Do you understand why I asked you to create a script?

**Comment by yashk2000** [comment]
Do I have to create a separate script to be run on fdroid?

**Comment by iamareebjamal** [comment]
Why else would I comment it on issue based on F-Droid release?

**Comment by iamareebjamal** [comment]
This has nothing to do with Travis, however, that single script will run on both travis and fdroid.


It will not have unnecessary things like sudo apt install and whatnot. Those utilities are already intalled in the environment

**Comment by yashk2000** [comment]
Sorry for misunderstanding it. I'll update this pr accordingly right now.

**Comment by iamareebjamal** [comment]
Simply make a script like this

https://github.com/iamareebjamal/opencv-android/blob/master/setup_opencv.sh

**Comment by yashk2000** [comment]
Ok.

**Comment by yashk2000** [comment]
@iamareebjamal the current script in this pr is alright? I just have to make a separate script using the same code, right?

**Comment by iamareebjamal** [comment]
> the current script in this pr is alright

No, it is not

It depends on Travis. It needs to be run in fdroid as well. See the script I linked which does not have any unnecessary dependency on any environment

File: phimpme.sh [hunk 1]
```diff
@@ -0,0 +1,101 @@
+language: android
```
Comment (iamareebjamal) [inline phimpme.sh]: Are you kidding me? What is this?

File: phimpme.sh [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    - android-$ANDROID_API_LEVEL
+    # For Google APIs
+    - addon-google_apis-google-$ANDROID_API_LEVEL
+    # Google Play Services
+    - extra-google-google_play_services
+    # Support library
+    - extra-android-support
+    # Latest artifacts in local repository
+    - extra-google-m2repository
+    - extra-android-m2repository
```
Comment (abishekvashok) [inline phimpme.sh]: looks more like a yml file than a .sh file

**Comment by yashk2000** [comment]
@iamareebjamal @abishekvashok how do I create the android environment in the bash script to run the build?

**Comment by iamareebjamal** [comment]
Just combine the 2 scripts to download and copy OpenCV SDK. It's not rocket science.

**Comment by iamareebjamal** [comment]
> Simply make a script like this
> 
> [iamareebjamal/opencv-android:setup_opencv.sh@`master`](https://github.com/iamareebjamal/opencv-android/blob/master/setup_opencv.sh) [](/iamareebjamal/opencv-android/blob/HEAD@{2019-07-27T16:25:14Z}/setup_opencv.sh)

**Comment by yashk2000** [comment]
@iamareebjamal ok. I only have to combine the openCV copy and downloading scripts. I thought that I had to convert the entire travis.yml to a bash script.

**Comment by yashk2000** [comment]
@iamareebjamal I have a doubt. This new script that is being made, in this, we only download openCV and copy the needed files to their respective destination folders?

**Comment by iamareebjamal** [comment]
YES!

**Comment by yashk2000** [comment]
@iamareebjamal ok. But I do not get how is it better than the way we are doing things now. We have to use another script besides travis in either case.

**Comment by iamareebjamal** [comment]
> This has nothing to do with Travis, however, that single script will run on both travis and fdroid.
> 
> It will not have unnecessary things like sudo apt install and whatnot. Those utilities are already intalled in the environment



> > the current script in this pr is alright
> 
> No, it is not
> 
> It depends on Travis. It needs to be run in fdroid as well. See the script I linked which does not have any unnecessary dependency on any environment

**Comment by iamareebjamal** [comment]
> But I do not get how is it better than the way we are doing things now. We have to use another script besides travis in either case.

No, you'll use the same single script in both travis and fdroid

File: phimpme.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
+mkdir $HOME/openCV/
+cd $HOME/openCV
```
Comment (iamareebjamal) [inline phimpme.sh]: You should not use $HOME as told before

File: phimpme.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+wget https://github.com/opencv/opencv/releases/download/4.0.1/opencv-4.0.1-android-sdk.zip
+
+unzip opencv-4.0.1-android-sdk.zip
+rm -rf opencv-4.0.1-android-sdk.zip
+# Rename folder name to opencv
+mv OpenCV-android-sdk opencv
+
+cd $TRAVIS_BUILD_DIR
```
Comment (iamareebjamal) [inline phimpme.sh]: You should absolutely have no dependency on TRAVIS at all. That is the entire point of this issue. To create a script which downloads and sets up OpenCV WITHOUT any environment dependency on Travis

The comment literally says, should work in both Travis and Fdroid. How can it work in fdroid if it depends on TRAVIS_BUILD_DIR

Comment (yashk2000) [reply phimpme.sh]: @iamareebjamal to provide the path to phimpme-android, I need to provide the travis path. How else can I copy the openCV files?

File: phimpme.sh [hunk 5]
```diff
@@ -1,20 +1,20 @@
-mkdir $HOME/openCV/
-cd $HOME/openCV
+mkdir openCV/
+cd openCV
```
Comment (iamareebjamal) [inline phimpme.sh]: Unnecessary

File: phimpme.sh [hunk 6]
```diff
@@ -1,20 +1,20 @@
...
 wget https://github.com/opencv/opencv/releases/download/4.0.1/opencv-4.0.1-android-sdk.zip
 
 unzip opencv-4.0.1-android-sdk.zip
 rm -rf opencv-4.0.1-android-sdk.zip
 # Rename folder name to opencv
 mv OpenCV-android-sdk opencv
 
-cd $TRAVIS_BUILD_DIR
+cd ..
```
Comment (iamareebjamal) [inline phimpme.sh]: Unnecessary

File: phimpme.sh (hunk 4 above)
Comment (iamareebjamal) [reply phimpme.sh]: You answered it yourself in next commit

File: phimpme.sh [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+wget https://github.com/opencv/opencv/releases/download/4.0.1/opencv-4.0.1-android-sdk.zip
+
+unzip opencv-4.0.1-android-sdk.zip
+rm -rf opencv-4.0.1-android-sdk.zip
+# Rename folder name to opencv
+mv OpenCV-android-sdk opencv
+
+cd ..
+
+mkdir phimpme-android/app/src/main/3rdparty
```
Comment (iamareebjamal) [inline phimpme.sh]:
```suggestion
mkdir papp/src/main/3rdparty
```

File: phimpme.sh [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+mkdir phimpme-android/app/src/main/jniLibs
+mkdir phimpme-android/app/src/main/staticlibs
+mkdir phimpme-android/app/src/main/jni/include
+
+mv openCV/opencv/sdk/native/3rdparty/* phimpme-android/app/src/main/3rdparty
```
Comment (iamareebjamal) [inline phimpme.sh]:
```suggestion
mv openCV/opencv/sdk/native/3rdparty/* app/src/main/3rdparty
```

**Comment by yashk2000** [comment]
@iamareebjamal thanks for the guidance. Is the script fine now?

File: .travis.yml [hunk 9]
```diff
@@ -67,7 +67,6 @@ licenses:
   - 'google-gdk-license-.+'
 
 before_script:
-#  - ./scripts/download_open_cv.sh
   - ./phimpme.sh
```
Comment (iamareebjamal) [inline .travis.yml]: Rename to setup_opencv.sh and use cache like the script in my link

**Comment by iamareebjamal** [comment]
Use version 4.1.1 and like I do in the linked script

**Comment by yashk2000** [comment]
> Use version 4.1.1 and like I do in the linked script

Then we'll have to update the openCV module being used currently too.

**Comment by yashk2000** [comment]
@iamareebjamal the script is updated. Please tell me about the openCV version to be used.

File: setup_opencv.sh [hunk 10]
```diff
@@ -0,0 +1,20 @@
+opencv_version="4.0.1"
+opencv_sdk_zip="cache/opencv-android-sdk.zip"
+download_url="https://github.com/opencv/opencv/releases/download/${opencv_version}/opencv-${opencv_version}-android-sdk.zip"
+mkdir cache/
```
Comment (iamareebjamal) [inline setup_opencv.sh]:
```suggestion
mkdir -p cache/
```

File: setup_opencv.sh [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+wget ${download_url} -O cache/opencv-android-sdk.zip
```
Comment (iamareebjamal) [inline setup_opencv.sh]: Where's -c?

File: setup_opencv.sh [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+unzip -qqo ${opencv_sdk_zip} -d opencv_sdk
+mkdir opencv/
+cp -r opencv_sdk/OpenCV-android-sdk/sdk/* opencv/
+rm -rf opencv_sdk
```
Comment (iamareebjamal) [inline setup_opencv.sh]: Why these extra unnecessary steps?

**Comment by yashk2000** [comment]
@iamareebjamal please review.

File: setup_opencv.sh [hunk 13]
```diff
@@ -0,0 +1,17 @@
+opencv_version="4.0.1"
+opencv_sdk_zip="cache/opencv-android-sdk.zip"
+download_url="https://github.com/opencv/opencv/releases/download/${opencv_version}/opencv-${opencv_version}-android-sdk.zip"
+mkdir -p cache/
+wget ${download_url} -c -O cache/opencv-android-sdk.zip
+
+unzip -qqo ${opencv_sdk_zip} -d opencv_sdk
```
Comment (iamareebjamal) [inline setup_opencv.sh]:
```suggestion
unzip -qqo ${opencv_sdk_zip}
```

File: setup_opencv.sh [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+
+mkdir app/src/main/3rdparty
+mkdir app/src/main/jniLibs
+mkdir app/src/main/staticlibs
+mkdir app/src/main/jni/include
+
+mv opencv_sdk/OpenCV-android-sdk/sdk/native/3rdparty/* app/src/main/3rdparty
```
Comment (iamareebjamal) [inline setup_opencv.sh]:
```suggestion
mv OpenCV-android-sdk/sdk/native/3rdparty/* app/src/main/3rdparty
```

As it was before without any extra useless folders

**Comment by iamareebjamal** [comment]
Add cache folder to travis cache

**Comment by iamareebjamal** [comment]
https://github.com/fossasia/phimpme-android/pull/2840#issuecomment-516859753

**Comment by iamareebjamal** [comment]
Also, move the script to scripts folder where it belongs with the rest of the scripts

**Pull request merged**